orchestrator-salt/src/test/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestratorTest.java

```diff
@@ -592,6 +601,101 @@ public void testInstallFreeIpaHaRepairTwoInstance() throws Exception {
         assertTrue(target3.contains("new_replica2.example.com"));
     }
 
+    @Test
+    public void testExistingIsPrimaryReplacement() throws Exception {
+        ObjectMapper mapper = new ObjectMapper();
+
+        GatewayConfig replica1 = mock(GatewayConfig.class);
+        GatewayConfig replica2 = mock(GatewayConfig.class);
+        Node replicaNode1 = mock(Node.class);
+        Node replicaNode2 = mock(Node.class);
+
+        when(replica1.getHostname()).thenReturn("replica1.domain");
+        when(replica2.getHostname()).thenReturn("replica2.domain");
+        when(replicaNode1.getHostname()).thenReturn("replica1.domain");
+        when(replicaNode2.getHostname()).thenReturn("replica2.domain");
+
+        PowerMockito.mockStatic(SaltStates.class);
+        HostAndRoleTarget replicaTarget = new HostAndRoleTarget("freeipa_replica", Set.of("replica1.domain", "replica2.domain"));
+        ArrayNode replicaRole = mapper.createArrayNode();
+        replicaRole.add("freeipa_replica");
+        PowerMockito.when(SaltStates.getGrains(eq(saltConnector), eq(replicaTarget), eq("roles")))
+                .thenReturn(Map.of("replica1.domain", replicaRole,
+                        "replica2.domain", replicaRole));
+
+        saltOrchestrator.installFreeIpa(replica1, List.of(replica1, replica2), Set.of(replicaNode1, replicaNode2), exitCriteriaModel);
+
+        ArgumentCaptor<ModifyGrainBase> modifyGrainCaptor = ArgumentCaptor.forClass(ModifyGrainBase.class);
+        verify(saltCommandRunner, times(2)).runModifyGrainCommand(eq(saltConnector), modifyGrainCaptor.capture(), eq(exitCriteriaModel), eq(exitCriteria));
+        List<ModifyGrainBase> modifyGrains = modifyGrainCaptor.getAllValues();
+        GrainAddRunner addReplacementRole = (GrainAddRunner) modifyGrains.get(0);
+        assertEquals("freeipa_primary_replacement", addReplacementRole.getValue());
+        assertEquals(Set.of("replica1.domain"), addReplacementRole.getTargetHostnames());
+        GrainRemoveRunner removeReplicaRole = (GrainRemoveRunner) modifyGrains.get(1);
+        assertEquals("freeipa_replica", removeReplicaRole.getValue());
+        assertEquals(Set.of("replica1.domain"), removeReplicaRole.getTargetHostnames());
+
+        ArgumentCaptor<SaltJobIdTracker> saltJobIdTrackerCaptor = ArgumentCaptor.forClass(SaltJobIdTracker.class);
+        verify(saltRunner, times(3)).runner(saltJobIdTrackerCaptor.capture(), eq(exitCriteria), eq(exitCriteriaModel), anyInt(), anyBoolean());
+        List<SaltJobIdTracker> saltJobIdTrackers = saltJobIdTrackerCaptor.getAllValues();
+        HighStateRunner highStateRunner = (HighStateRunner) saltJobIdTrackers.get(0).getSaltJobRunner();
+        assertTrue(highStateRunner.getTargetHostnames().contains("replica1.domain")
+                || highStateRunner.getTargetHostnames().contains("replica2.domain"));
+        assertEquals(1, highStateRunner.getTargetHostnames().size());
```

Comment:
Just a suggestion, you could also use
assertThat(collection, hasSize(x));

Reply:
thanks, I'll keep that in mind next time